### PR TITLE
Task 12.12	Comment on a review (user signed in)

### PR DESCRIPTION
**Pull request by oDoHongSon**
Task 12.12	Comment on a review (user signed in)

### CODE REVIEW

File: app/controllers/review/comments_controller.rb [hunk 1]
```diff
@@ -1,5 +1,29 @@
...
+    if @comment.save
+      @review = @review.reload.decorate
+      @comments = @review.comments.earlier_created.includes(:user, :branch).decorate
+    else
+      flash.now[:failed] = @comment.errors.full_messages.first
+    end
+  end
+
+  private
+  def comment_params
```
Comment (lucnv) [inline app/controllers/review/comments_controller.rb]: phan param nay chi lay param tu request, ko merge o day

File: app/controllers/review/comments_controller.rb [hunk 2]
```diff
@@ -1,5 +1,29 @@
...
+  def new
+    @new_comment = Comment.new
+  end
+
   def index
     @comments = @review.comments.earlier_created.includes(:user, :branch).decorate
   end
+
+  def create
+    if current_user.working_center.present? && current_user.branches_under_management.include?(@review.branch)
```
Comment (lucnv) [inline app/controllers/review/comments_controller.rb]: neu phuc tap qua thi  dung service

File: app/controllers/review/comments_controller.rb (hunk 1 above)
Comment (oDoHongSon) [reply app/controllers/review/comments_controller.rb]: edited

File: app/models/comment.rb [hunk 3]
```diff
@@ -8,4 +8,6 @@ class Comment < ApplicationRecord
   acts_as_paranoid
 
   counter_culture :review
+
+  PARAMS = [:content, :user_id, :branch_id, :review_id]
```
Comment (lucnv) [inline app/models/comment.rb]: chi lay branch_id cho center_comment

File: app/controllers/review/comments_controller.rb [hunk 4]
```diff
@@ -1,5 +1,28 @@
...
+  def new
+    @new_comment = Comment.new
+  end
+
   def index
     @comments = @review.comments.earlier_created.includes(:user, :branch).decorate
   end
+
+  def create
+    if current_user.manage_branch? @review.branch
```
Comment (lucnv) [inline app/controllers/review/comments_controller.rb]: tach doan nay ra1 method

File: app/models/user.rb [hunk 5]
```diff
@@ -42,6 +42,10 @@ def branches_under_management
     end
   end
 
+  def manage_branch? branch
+    working_center && branches_under_management.include?(branch)
```
Comment (lucnv) [inline app/models/user.rb]: branches_under_management.try :include?, branch

File: app/models/comment.rb [hunk 6]
```diff
@@ -8,4 +8,9 @@ class Comment < ApplicationRecord
   acts_as_paranoid
 
   counter_culture :review
+
+  PARAMS = [:content, :user_id, :review_id]
```
Comment (lucnv) [inline app/models/comment.rb]: trong scope nay thi ko can user_id, review_id
them nuea e luu y la neu permit user_id thi  nguoi dung co the tao comment voi danh nghia nguoi dung khac

File: app/models/comment.rb [hunk 7]
```diff
@@ -8,4 +8,9 @@ class Comment < ApplicationRecord
...
+
+  validates :content, presence: true, length: {minimum: Settings.validations.comment.content.min_length,
+    maximum: Settings.validations.comment.content.max_length}
```
Comment (lucnv) [inline app/models/comment.rb]: allow_blank: true (trong lenght: {})

File: app/controllers/review/comments_controller.rb [hunk 8]
```diff
@@ -1,5 +1,32 @@
 class Review::CommentsController < Review::BaseController
+  before_action :build_comment, only: :create
```
Comment (lucnv) [inline app/controllers/review/comments_controller.rb]: ham nay ko phai thuc hien before_action. Viet truc tiep vao action create

File: app/controllers/review/comments_controller.rb [hunk 9]
```diff
@@ -1,5 +1,32 @@
...
+  end
+
+  private
+  def comment_params
+    params.permit Comment::PARAMS
+  end
+
+  def build_comment
+    if current_user.manage_branch? @review.branch
+      @comment = CenterComment.new(comment_params. merge branch_id: @review.branch_id).decorate
```
Comment (lucnv) [inline app/controllers/review/comments_controller.rb]: build comment thi chi build ra 1 object ActiveRecord thoi, ko decorate o day

File: app/controllers/review/comments_controller.rb [hunk 10]
```diff
@@ -1,5 +1,33 @@
...
+    @new_comment = Comment.new
+  end
+
   def index
     @comments = @review.comments.earlier_created.includes(:user, :branch).decorate
   end
+
+  def create
+    build_comment
+    @comment = @comment.decorate
```
Comment (lucnv) [inline app/controllers/review/comments_controller.rb]: @comment = build_comment.decorate

File: app/controllers/review/comments_controller.rb [hunk 11]
```diff
@@ -1,5 +1,33 @@
...
+  end
+
+  private
+  def comment_params
+    params.permit Comment::PARAMS
+  end
+
+  def build_comment
+    if current_user.manage_branch? @review.branch
+      @comment = current_user.comments.build comment_params.merge(branch: review.branch,
```
Comment (lucnv) [inline app/controllers/review/comments_controller.rb]: delete `@comment = `

File: app/controllers/review/comments_controller.rb [hunk 12]
```diff
@@ -1,5 +1,33 @@
...
+        type: CenterComment.name, review: review)
+    else
+      @comment = current_user.comments.build comment_params.merge type: UserComment.name, review: review
```
Comment (lucnv) [inline app/controllers/review/comments_controller.rb]: current_user.user_comments. Them user_comment vao association cua user

File: app/decorators/comment_decorator.rb [hunk 13]
```diff
@@ -21,7 +21,7 @@ def owner_name
     if object.is_a? UserComment
       user.full_name
     else
-      branch.name
+      branch.center_name + " - " + branch.name
```
Comment (lucnv) [inline app/decorators/comment_decorator.rb]: branch.full_name

File: app/controllers/review/comments_controller.rb [hunk 14]
```diff
@@ -1,5 +1,31 @@
...
   def index
     @comments = @review.comments.earlier_created.includes(:user, :branch).decorate
   end
+
+  def create
+    @comment = build_comment.decorate
+    if @comment.save
+      @review = @review.reload.decorate
+    else
+      flash.now[:failed] = @comment.errors.full_messages.try :first
```
Comment (lucnv) [inline app/controllers/review/comments_controller.rb]: da fail thi chac chan co errors. Doan nay tot nhat show het errors ra

File: app/controllers/review/comments_controller.rb [hunk 15]
```diff
@@ -1,5 +1,31 @@
 class Review::CommentsController < Review::BaseController
+  def new
```
Comment (lucnv) [inline app/controllers/review/comments_controller.rb]: authen server side?

File: app/controllers/review/comments_controller.rb [hunk 16]
```diff
@@ -1,5 +1,37 @@
 class Review::CommentsController < Review::BaseController
+  before_action :authorize_user, only: :create
```
Comment (lucnv) [inline app/controllers/review/comments_controller.rb]: tai sao ko authen cho method new?

File: app/views/review/comments/new.js.erb [hunk 17]
```diff
@@ -0,0 +1,3 @@
+$("#review-details-<%= @review.id %> .new-comment-form")
+  .html("<%= j (render 'shared/comments/new_comment', review: @review) %>");
+$("#review-details-<%= @review.id %> .new-comment").addClass("hidden");
```
Comment (lucnv) [inline app/views/review/comments/new.js.erb]: hide()

**Pull request merged**